packages/scrapers-v2/src/lexer/lexer.ts

```diff
@@ -0,0 +1,432 @@
+  const hour = parseHour(c2);
+  return { hour, description, type };
+};
+
+const constructPlainCourseRow = (
+  $: CheerioStatic,
+  tr: CheerioElement
+): CourseRow<HRowType.PLAIN_COURSE> => {
+  const [code, desc, hourCol] = ensureLength(3, tr.children).map($);
+  const { subject, classId } = parseCourseTitle(parseText(code));
+  const description = parseText(desc);
+  const hour = parseHour(hourCol);
+  return { hour, description, type: HRowType.PLAIN_COURSE, subject, classId };
+};
+
+const constructOrCourseRow = (
+  $: CheerioStatic,
+  tr: CheerioElement
+): CourseRow<HRowType.OR_COURSE> => {
+  const [code, desc] = ensureLength(2, tr.children).map($);
+  // remove "or "
+  const { subject, classId } = parseCourseTitle(
+    parseText(code).substring(3).trim()
+  );
+  const description = parseText(desc);
+  // there may be multiple courses in the OR, so we can't backtrack
+  return { hour: 0, description, type: HRowType.OR_COURSE, subject, classId };
+};
+
+const constructMultiCourseRow = (
+  $: CheerioStatic,
+  tr: CheerioElement
+): MultiCourseRow<HRowType.AND_COURSE> => {
+  const [code, desc, hourCol] = ensureLength(3, tr.children).map($);
+  const titles = code
+    .find(".code")
+    .toArray()
+    .map($)
+    .map(parseText)
+    .map(parseCourseTitle);
+  const firstDescription = parseText(desc.contents().first());
+  const restDescriptions = desc
+    .children(".blockindent")
+    .toArray()
+    // ignore the first four characters, "and "
+    .map((c) => parseText($(c)).substring(4).trim());
+  const descriptions = [firstDescription, ...restDescriptions];
+  if (titles.length !== descriptions.length) {
+    const msg = `found titles: ${titles.length} !== found descs: ${descriptions.length}`;
+    throw new Error(msg + titles + descriptions);
+  }
+  const courses = titles.map(({ subject, classId }, i) => ({
+    subject,
+    classId,
+    description: descriptions[i],
+  }));
+  const hour = parseHour(hourCol);
+  return {
+    hour,
+    type: HRowType.AND_COURSE,
+    description: descriptions.join(" and "),
+    courses,
+  };
+};
+
+const constructRangeLowerBoundedMaybeExceptions = (
+  $: CheerioStatic,
+  tr: CheerioElement
+):
+  | WithExceptions<
+      RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS>
+    >
+  | RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED> => {
+  const [desc, hourCol] = ensureLength(2, tr.children).map($);
+  const hour = parseHour(hourCol);
+  // text should match one of the following:
+  // - CS 9999 or higher[, except CS 9999, CS 9999, CS 3999,... <etc>]
+  // - Select from any HIST course numbered 3000 or above.
+  // - Complete three HIST courses numbered 2303 or above. Cluster is subject to Department approval.
+  const text = parseText(desc);
+  // should match the form [["CS 9999", "CS", "9999"], [...]]
+  const matches = Array.from(text.matchAll(RANGE_LOWER_BOUNDED_PARSE));
+  const [[, subject, , , , id], ...exceptions] = ensureLengthAtLeast(
```

Comment:
oh these commas just mean like the position of the array. So like if I have array of `[["CS 9999", "CS", "9999"], ["HI 9999", "HI", "9999"], ["WE 9999", "WE", "9999"]]`, then if I do `const [[, subject]] = matches`, it will take the second element of the first array as `subject`, which would be "CS"

Reply:
yeaH this javascript syntax is kinda wack... but ya what cindy said